include/storage/flash_map.h

```diff
@@ -247,6 +247,21 @@ const struct device *flash_area_get_device(const struct flash_area *fa);
  */
 uint8_t flash_area_erased_val(const struct flash_area *fa);
 
+/**
+ * Reads len bytes from off, and checks if the read data is erased.
+ *
+ * @param[in]  fa  Flash area
+ * @param[in]  off Offset relative from beginning of flash area to read
+ * @param[out] dst Buffer to store read data
+ * @param[in]  len Number of bytes to read
```

Comment:
Should document the constraint on valid lengths here:

    len Number of bytes to read.  The value must be positive.

Though honestly I don't like the name of this function *or* its API, relative to what it does.  Since we're refactoring, can we correct the API: 

    int flash_area_try_read(const struct flash_area *fa, off_t off,
￼			     void *dst, size_t len, bool* is_erased);

?  Then we don't have a really non-standard behavior when `len == 0` that would make callers have to check the value.

I don't really like that name either; but I understand the role of this to be the same as a normal read, but to also indicate whether what was read was erased.

Maybe we shouldn't add this, but instead have:

    bool flash_area_erased(const struct flash_area *fa, const void *data, size_t len);

that returns `true` or `false`, with `false` if somebody passes `NULL` .

Yeah, I think that makes more sense.  It can even be a helper function that doesn't need to get implemented in each driver.

Reply:
Corrected badly chosen original name to "flash_area_*data*_is_erased".  It could also be "flash_area_buffer_is_erased".

Reply:
Those names where not created here, they come from MCUboot's API requirements.

Reply:
> Though honestly I don't like the name of this function _or_ its API, relative to what it does. Since we're refactoring, can we correct the API:
> 
> ```
> int flash_area_try_read(const struct flash_area *fa, off_t off, void *dst, size_t len, bool* is_erased);
> ```
> 

This function's name does not tell me that it's is checking for an erased block, although I can see by the last parameter's name...

> 
> ```
> bool flash_area_data_is_erased(const struct flash_area *fa, const void *data, size_t len);
> ```
> 
> that returns `true` or `false`, with `false` if somebody passes `NULL` .
> 
> Yeah, I think that makes more sense. It can even be a helper function that doesn't need to get implemented in each driver.

AFAIR the original idea was to have a way of read and check that it might be erased in a single read, so that users of the API don't need to manually to the erased checking. If I understand correctly this API would require two reads.

Reply:
No, that wasn't the intent.  Every implementation of `flash_area_read_is_empty` is basically:
* invoke the standard flash read
* iterate over the result looking for a non-erased value.

What I'm proposing with the badly named `flash_area_data_is_erased` is to make a function that does the second half, given the data that was read in the first half.  The only reason this is even in the flash area API is that we need an `fa` parameter is to find out what erase value we're checking against.  There's no I/O involved.

This would be generically useful, e.g. if you only want to check whether a certain sub-sequence of what you read was erased, so it makes a some sense to add it to Zephyr's API.

Then for applications that require `flash_area_data_is_erased` it's just a call to `flash_area_read` followed by a call to the function that checks it's all erased.

Reply:
Ok, I like that.

Reply:
this function is not in scope anymore